### DISCUSSION TITLE
Bad Performance in Pull Request File Tree

### DISCUSSION THREAD

**Discussion by joshuambg** [community/community#39341] [Pull Requests]
### Select Topic Area

Bug

### Body

This is a topic that has been documented in the community postings. I can't find anyone from GitHub that answers.
https://github.com/community/community/discussions/10830
https://github.com/community/community/discussions/33663

Large PRs with lots of files crash GitHub PRs. This is a real problem that needs to get on GitHub's planning board. Use the latest version of any browser. The problem is with the GitHub feature.

- Create a PR in GitHub with lots of files
- Try using the PR UI and PR file tree

Expected
- Fast and responsive UI

Result
- Browser tab can crash
- Browser freezes
- Scrolling clunky
- Chrome warns about slow unresponsive page

**Comment by borbudak-jadbio** [comment]
Any updates on this? 

Even with only a couple hundred changed files Github code review can be extremely slow, to the point of it being almost unusable. The file tree search feature is also unusable. It takes ~10 seconds just for the cursor to appear once clicked, and typing/searching is even worse.

**Reply by themanciraptor** [reply]
As a workaround, you can use Firefox. Using Firefox, my old intel mac (6c, 16GB ram) renders large pr's pretty much lag-free while a few of my coworkers' apple M2's running chrome are freezing and lagging like crazy.

**Comment by CoenWarmer** [comment]
If you have more than 3k changes the UI just informs you it won't show you the remainder. C'mon Github, can't we do better than that?

**Comment by EdmondsZinc** [comment]
Still nothing on this one? How is Chrome being left out here? 
I can confirm FireFox does appear to be a valid workaround.  I just tested my PR and 2000 files is unusable in chrome but works fine in Firefox... womp. 
You can't even expand the file tree width to see longer file names... pretty sad.

**Reply by alexandru-pap-osf** [reply]
Firefox works. Currently i have 2k files in a single PR, and I could open it so fast in Firefox.
Thanks for sharing the info.

**Comment by andreieftimie** [comment]
I used to switch from Safari to Chrome for a _small_ performance boost on huge PR's.

Today Chrome is also choking. I am literally unable to review large PR's :(

**Comment by udlose** [comment]
I am also having this problem. We had a refactor that could only be done in a single PR w/o breaking the code in the repo.  We had ~2,200 files changed in the PR.  Why isn't there some sort of paging that exists?

**Comment by mpiggott** [comment]
Even a couple hundred files renders the UI extremely slow - filtering a tree with 200 entries shouldn't require locking up the browser for 10-seconds.

**Comment by crobinson-expl** [comment]
I'm having the same issue over the last week.  A PR with just 307 files has taken me days to go through.  Just marking a file as viewed takes around 5 seconds.  If I have to click to load the diff (on files down past some file count threshold, even though they are small), then that's another 8-10 seconds.  I have reviewed much larger PRs in the past that have not had these performance issues.

**Reply by george-of-croton** [reply]
i agree. I have an 800 hundred liner im trying to review and im getting multi-second lag on text input when writing review comments.

**Reply by jeremy-hunter** [reply]
For whatever reason Firefox handles reviewing a 3000 file PR just fine while chrome and safari are unusable.
Chrome says the memory heap for the PR on first load is 66MB, Firefox has the memory heap at ~450MB on load - maybe it's just a browser trying to conserve memory

**Reply by crobinson-expl** [reply]
A team member who uses the Brave browser let me know that they aren't experiencing the same issues, so I tried it out.  Marking files as "viewed" is much snappier, but the time to being able to type in anything on a code comment seems about the same.

I noticed in the browser developer tools network tab that Edge makes many posts to https://api.github.com/_private/browser/stats whereas Brave doesn't do this (Brave left, Edge right): 
![image](https://github.com/community/community/assets/84980421/f77591fb-9c4e-4fa3-b500-98eb0de23bbd)

I opened a ticket with GitHub Support on Friday but haven't received any update yet.

**Reply by crobinson-expl** [reply]
@jeremy-hunter Thanks for the Firefox info!  I tried it there and it works great, even better than it does in Brave.  Everything is responsive and usable like it used to be in Edge.

**Comment by milivojsa** [comment]
I can confirm that Firefox works great. Here's my pull request:
![image](https://github.com/community/community/assets/17594274/23cc9b1f-3276-423f-a41a-9645cfeb7c5f)

Chrome was stuck when I switched to the Files changed section while Firefox works like a charm.

I compared memory usage. Chrome comes up to 2GB, while Firefox is most of the time 0.7 - 1GB on the same page.

**Comment by zgavin1** [comment]
I have a pretty mild PR compared to some of these examples (1200 files, 1700 changes) and Chrome outright crashes, and firefox seems barely any better if at all.

**Comment by Blafasel3** [comment]
Feels like this gets worse with every Github update. Please fix this, this has the potential to make people move to other tools.

**Comment by ValentinGurkov** [comment]
I've been using mostly Vivaldi except when I need to review PRs because it lags there. For viewing PRs I use Chrome but it's lagging now there too.

The PR is not that big, 3,700 additions, 1800 removals. I've worked with much larger PRs without lag in the past.

**Comment by Sinarab** [comment]
I'm having the same issue here using Chrome. I tried using Firefox as a temporary solution to this lag situation, and it works only partially. Marking items as 'viewed' is fine, but searching for comments through the 'jump to conversation' feature doesn't work on both crashing the app.

Hope someone can solve this issue asap. :crossed_fingers:

**Comment by Jakoban2411** [comment]
Still a huge issue. I've been using Brave and for PRs with >1000 files, despite filtering out to have only ~100 files, there's a huge lag, the Ui is unresponsive, Brave keeps asking me to kill the tab.

**Comment by paul-vd** [comment]
I can also confirm tested today with a PR of 2600+ it was snappy on firebox, but really slow on chrome.

on FF the only thing a little slow is when adding a comment to a line of code, but still much better than chrome.

**Comment by vmendiola** [comment]
Just experienced this today with a PR having ~800 files changed. Sadly, it is unbearable to work with.

**Comment by nlvedwin** [comment]
Same issue here it's super slow and making everything laggy. If you use firefox to view the changes it works perfectly fine

**Comment by joaopedrodcf** [comment]
The experience in web with PRs with lot of changes >300 it's terrible! The issue is that the all list is being loaded, at least in mobile we do have lazy loading there that does help with performance

**Comment by ovargas** [comment]
I have a merge request with 392 changed files. To try to improve performance, I’ve added the following patterns to my .gitattributes file to disable diffs:

```text
vendor/** -diff
protobuf/** -diff
**/mocks/** -diff
```

While the UI no longer displays diffs for these files, it is still slow to load. It would be nice to:
- Automatically mark all files matching these patterns as Viewed, or have a way to do it in bulk.
- Have an option to exclude those files from the list entirely so I can focus on the remaining files in the PR.
- Improve the navigation and the loading per folder or similar.

**Comment by clicktodev** [comment]
I really wish this part of github was open source, atleast then motivated independant devs can analyze the problem and push a fix by themselves.

**Comment by ChristianOellers** [comment]
Having the same issue in MacOS Chrome 131. For about 500 changed files, it's impossible to open even a Pull Request. Plus, just opening one, I would not expect or need an immediate diff to be visible.

Some ideas for an optimized UI:

1. Can't we just remove the auto-preview of the diff and make it opt-in or available on toggle?
2. If the displaying logic is inverted, I can also imagine the 'filter by filetype' tool, that really optimizes the performance, to work in the opposite way -> Toggle on what to see, not off what to remove

This really would need a change – The web UI is completely useless for large changes. Given the logical structuring of the project, making smaller Pull Requests is not an option, and large changes are a common requirement.

🙏

**Comment by JulianKingman** [comment]
376-file PR, mutiple second delay in Safari. Firefox is _reasonably_ snappy, but I still get blank spots when scrolling where it takes it a minute to catch up. Also I don't think it's unreasonable to expect all the file names to load immediately and be searchable, regardless of the size. We're not talking about that much data here, are we? Even a big PR is probably in the 10's of megabytes...

**Comment by zcoffin86** [comment]
Same issue here. 54 Files changed, +640 -621. Chrome is unbearably slow, but Firefox has no issue. Looking forward to Chrome being usable again.

**Comment by henrahmagix** [comment]
I'm seeing a 2-3 second delay in Safari 18.1.1 on macOS 15.1.1, when clicking a line to add a comment:
- with the network panel open, I click the blue plus button and it's 1-2 seconds until a request to `/review_comment/suggestion_button` is made, which returns quite quickly with the html for the suggestion button: `Add a suggestion, <Ctrl+g>`
- then it's another 1-2 seconds to render the comment box

And it's the same total delay when clicking cancel on the comment box.

This is on a PR with 19 files changed, 465 lines added, 272 lines removed, so not very big at all.

I've tried quitting and reopening Safari, I've tried restarting, and I've tried refreshing without the cache: it's still super slow 😭

**Reply by henrahmagix** [reply]
I'm also seeing a 2 second delay when clicking "Start a review" until it becomes disabled, and then another 2-3 seconds until it submits the comment. It literally feels like it's running on a `setTimeout(..., 1000)` call or something 🤔

**Comment by dlohvinov** [comment]
Same here, in Safari, Chrome, and Firefox too. 🥲

**Comment by dungkhoai** [comment]
same, all repo

**Comment by ppamorim** [comment]
I am unable to create PRs due to this issue, we have automated actions to modify source code and those apply massive changes (think of a linter). I tried on both Safari and Firefox and both are unable to let me open the PR. The page freezes. I will need to force merge the PR instead.

**Comment by bmanturner** [comment]
Unable to open a pull request at all anymore. Before it took a few tries, but now the entire page just freezes, doesn't matter what browser.

**Comment by TKTang1** [comment]
When reviewing a large PR, GitHub is massively slow and unusable. Azure DevOps surprisingly has a UI/UX that's actually performantly flawless when reviewing a PR of identicial size.

**Comment by nbolton** [comment]
Above 50 files changed the UI becomes unusable. How about paging the results instead of loading more than the browser can handle?

Files changed: 681
Browser: Chrome (with extensions)

**Reply by nbolton** [reply]
I think I have "solved" this. Using a fresh Firefox with no extensions, it is fairly responsive.

**Reply by zcoffin86** [reply]
The issue persists for me when running chrome with --disable-extensions.

**Reply by nbolton** [reply]
Oh, what about if you use Firefox? Would be good to narrow this down if browser specific.

**Comment by ValentinGurkov** [comment]
I've noticed that my NordPass browser extension can also affect the performance of the PR page.

**Reply by nbolton** [reply]
Ah, browser extensions. I should have realised.

**Reply by zcoffin86** [reply]
The issue persists for me when running chrome with --disable-extensions.

**Reply by nbolton** [reply]
> Oh, what about if you use Firefox? Would be good to narrow this down if browser specific.

**Comment by DanielGumpHTT** [comment]
Safari 18.2 on a Macbook Pro M1 hangs up for me when creating new PRs that contain only a couple hundred files changed. This started about a week ago.

**Reply by nbolton** [reply]
What happens if you use Firefox?

**Comment by nbolton** [comment]
@joshuambg Based on all discussiuons so far, I suspect this is a Chrome-specific issue. Everyone I asked so far finds that Firefox fixes the problem. If you agree, please rename this issue to: 

**"Bad performance in Pull Request file tab on Chrome"**

I believe this will help GitHub devs understand why the issue occurs and may help them fix it. It's likely most of the dev team uses Firefox, and so they don't see the issue very often.

**Reply by OddMathisenNOV** [reply]
This is not only affecting the Chrome browser, but seemingly any Chromium (Browser Codebase) related browsers.

**Reply by nbolton** [reply]
Yes, I should have said Chrome/Chromium. I also hear Safari is quite bad too. Someone else also mentioned Firefox is better but not perfect (though I don't think they disabled their add-ons).

**Reply by AntonMikhailovAH** [reply]
On Chrome issue isn't as bad as on Safari

**Comment by nbolton** [comment]
# Workaround

**Temporary solution until GitHub fixes this:** Use Firefox with _add-ons disabled_.

Many people have reported that Firefox (with add-ons disabled) has far better performance on the PR file tab compared to Chrome, Chromium and Safari. Many say that disabling extensions in Chrome doesn't seem to help (I can confirm this).

Re-post from: #33663

**Comment by ppamorim** [comment]
@joshuambg This gets bad when the Github compare screen defaults to the default repository branch, if you are working in the `development` branch and you try to open a PR, and if your `development` is massively different from `main`, your pre-PR screen will never finish loading.

**Comment by Manfred** [comment]
It's not really in my nature to comment on performance in products because I think developers spend enough time with their product not to notice changes. However, doing code reviews on the Files Changed tab is our primary interaction with GitHub and performance has gotten so bad in recent months that we started doing code reviews in Fork instead of on GitHub and are actively looking at hosted alternatives for our team.

**Comment by xvyslo05** [comment]
bump!

currently on Safari 18.1.1, github's both PRs and Projects are painfully slow. I literally want to shoot to my knee, instead of using it

**Comment by wasade** [comment]
What browser do GitHub devs use for PRs?

**Comment by arthurfiorette** [comment]
Please fix this out!! It's painfully slow on Brave v1.75.181 even for small PRs.

**Comment by ghost** [comment]
**My Workaround is:**
Use VSCode with Github PR extension.

And if one wants to do the PR in the browser:
https://vscode.dev/github/{username or company}/{repo}/pull/{prNumber}

**Reply by marianhlavac** [reply]
You don't have to build the URL manually, just press the dot <kbd>.</kbd> key on the PR page and you'll jump to the VSCode directly.

**Comment by pazoozooCH** [comment]
+1

This is in a desolate state for even medium sized PRs. Interestingly, it doesn't help much if I filter out the already _Viewed_ files. Somehow it remains slow, even if 90% of the files are not shown. Can't believe this has been an issue for over 2 years now... 🤯

**Comment by kevinfiol** [comment]
In the last few years, Github has really shat the bed on multiple features, either actively making them less useful or less performant. This one is unbelievably embarrassing; I can't believe this is unfixed almost 3 years later.

**Comment by ashton-seek** [comment]
Please look into this.

**Comment by bitaccesscomau** [comment]
This is absolutely pathetic.

**Comment by jonholdsworth** [comment]
Hello. I'm afraid this is still happening in April 2025.

**Comment by LucasMontoya4** [comment]
Why is there no quick workaround to this? And why isn't this problem more common? Surely every single person with a GitHub repo that has large PRs is having the same problem!?

**Reply by yuan-kuan** [reply]
Firefox is the workaround. I just refreshed my Firefox just for this and confirmed this workaround is still valid after a year. Safari is simply unusable.

**Comment by pyrtsa** [comment]
We're also getting bitten by this issue. The PR doesn't even need to be very big: you only need some 100 removed files in one PR for the checkbox to take 3+ seconds to click in macOS Safari; see screen recording further below.

I added some breakpoints around in Safari Web Inspector and found two lines modifying the DOM tree to be exceptionally slow, taking about a second for the debugger to step over to the next line:
- setting `checkbox.disabled = true` on `https://github.githubassets.com/assets/app/assets/modules/github/pulls/toggle-user-reviewed-file.ts` line 33, and
- replacing the form `target.replaceWith(response.html)` on line 54.

Maybe the DOM tree would need some pruning or optimisation?

----

https://github.com/user-attachments/assets/94b71c26-375c-4965-8513-3ef80da082d1

<img width="769" alt="toggle-user-reviewed-file" src="https://github.com/user-attachments/assets/c095080d-28d5-4b6f-9d94-970756d8e0d9" />

**Comment by marianhlavac** [comment]
While I hate to suggest this, I recommend contacting GitHub customer support directly about this issue, since it's clear they don't read the community discussions. Everyone experiencing this problem should reach out. Support will likely respond with a template message saying they don’t have an ETA and that they understand the reply isn’t ideal. Still, it’s worth expressing your dissatisfaction—especially if you're a paying customer. Be sure to emphasize that. Unfortunately, there aren’t many other options. Well you can always leave GitHub, but we all know that it's easy to say this, but hard to execute.

**Reply by msw-kialo** [reply]
I would suspect/hope they will link to https://github.com/github/roadmap/issues/1102; if the roadmap is up-to-date this issue should be at least soften within the next 2 months.

**Reply by embostan** [reply]
Every change they made in the past years made performance worse.

**Comment by GoodPie** [comment]
Once or twice a year, I have to deal with a large PR where lot's of files are removed or our icon package changes. Once or twice a year, I always end up in this same thread. It's disappointing that this hasn't been resolved yet.

**Comment by joshuambg** [comment]
Hiii. OP here. Pretty clear GitHub dev either does not read these threads or the message is not getting through.

If you happen to end up here, frustrated but motivated, one thing to try is contacting the dev team directly. Search for GitHub engineering titles on LinkedIn, then try to find their emails. If you or your company has a paid plan, then why are you even here? Complain through official channels.

Otherwise, the answer could very well be that GitHub management is fully aware of this problem. It would mean investing a lot of resources into refactoring the code for PRs. The answer is no and they know you're not switching platforms.

**Comment by saudsami** [comment]
Unbeleivable that this has still not been resolved.

**Comment by embostan** [comment]
I'm getting pretty convinced the GitHub teams do not use their own product

**Comment by udlose** [comment]
Maybe this will finally be fixed!

https://github.blog/changelog/2025-06-26-improved-pull-request-files-changed-experience-now-in-public-preview/

**Reply by importRyan** [reply]
Can't comment on perfection (used it only on macOS beta), but it remarkably improved performance reviewing a 8k line branch. Great! Sadly, "large diff" files are still defaulted to hidden and buried in between other more-than-screen-height diffs.

**Comment by vedmant** [comment]
Having the same problem, but I manage to do PR review in Firefox, it's slow but doesn't hang indefinitely like chrome.

**Comment by hasansino** [comment]
UI on chrome is very unstable at the moment.
Event now, when I am typing this message, I have crazy latency.

**Comment by nlvedwin** [comment]
Still an issue today. I have to use Firefox to smoothly check PRs

**Comment by kuteninja** [comment]
+1 on this, Safari is particularly painful on PRs with 100+ files

I think that a quick fix would be to have all the changes collapsed when the PR is over N files/changes long, and that might make it better.

Other ideas would be to have some dynamic pagination on the page which only shows some files and the rest are kept on a display hidden, might cause some scrolling bumps though when switching between them 🤔

**Comment by MunhozTheMage** [comment]
Our team is having this issue with PRs with only 40 changes, it's annoying to review anything in there.

**Comment by jonholdsworth** [comment]
Is there any way we can get this escalated?

**Comment by roonskipp** [comment]
Experiencing the same issue in safari. Incredibly unresponsive UI when reviewing PRs.

**Comment by inriverisac** [comment]
This is so bad. Unbelivable. We were planning from switching from ADO to GitHub but this is unusable. How're people using this service?

**Comment by EmileBV** [comment]
This is still a very real issue, there's around 5 seconds of delay when doing anything on a pull request with about 200 files for me on Firefox.

For what it's worth if you're using VSCode, the performance using the [Github Pull Requests plugin](https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github) is excellent and gives access to more tools while reviewing. I'll be switching to this method until they figure things out.

**Comment by dungkhoai** [comment]
Temporary solution: Do not use the “Create Auto Pull Request” button. Instead, the user should click “New Request”, then select the target branch and the compare branch to proceed.

**Comment by uniquelau** [comment]
Still an issue, yawn!

**Comment by akermabon** [comment]
Is there another known workaround than using a different browser ? The company I work at enforces Chrome usage :/

**Reply by EmileBV** [reply]
If you're using VS or VSCode you could try the plugin: [Github Pull Requests](https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github)

**Comment by jonholdsworth** [comment]
This is becoming bad enough I've started using Firefox for some github PR's, as Chrome variants cannot be relied upon to not exhibit this bug.

**Reply by jonholdsworth** [reply]
It's possible the 'New experience' view fixes this, as it seems performant in Chrome.

**Reply by imanmahjoubi** [reply]
@jonholdsworth are you still experience performance issues on the new file changed page?

**Reply by jonholdsworth** [reply]
@imanmahjoubi 
The page takes a while to load, but once it does it seems to work Ok!
It doesn't drag the CPU into the ground and I can add and delete comments.

On THIS page where I'm typing this now, the text input is still laggy, but it also is better than it was!

**Reply by jonholdsworth** [reply]
I take it back. The performance is still so bad on big PRs that I have to use Firefox.
On Chrome the response lags worse and worse until it becomes useless. The page does not consume 100% CPU though, that is the only improvement.

**Comment by jonholdsworth** [comment]
I've noticed this message start appearing, and also the performance may indeed have improved.

`This page has been optimized for large pull requests. Some browser features (like Find on Page or Select All) may not work as expected.`

**Comment by dwolrdcojp** [comment]
Still horrendous

**Comment by jonholdsworth** [comment]
I submitted a notice to Github support via their contact form, including the url of this thread and explaining the issue.
Feel free to pile on.
You need to sneak past their answer avoidance mechanisms and AI.

https://github.com/contact

**Reply by jonholdsworth** [reply]
Ok this is still not fixed and it's getting silly.
I'll see if I can follow-up on that ticket.

**Reply by jonholdsworth** [reply]
Nope, so I'm just going to raise another one.

**Reply by jonholdsworth** [reply]
Ticket created.

**Comment by jonholdsworth** [comment]
Maybe they've fixed it??

I had a large PR 2 days ago and it seemed to be reasonably reponsive.
Also this message box has started appearing 'This ... has been optimised for large PRs' - so microsoft _are_ aware of the issue!

Could others test and see if they are getting the same problem still?

--- ticket response below ---

## Please do not write below this line ##

Your request (Ticket 4386935) has been updated.

To review the status of your request and add additional comments, please visit: Ticket 4386935.

	

Rick (GitHub Support)

May 15, 2026, 7:27 PM UTC
Hi there jonholdsworth,
 
Sorry to hear you are experiencing issues when using GitHub.com. 
 
Based on the information you have provided, I was unable to replicate this issue using a chrome browser. Is there any chance you could have conflicting browser extensions that may be causing this behavior?
 
If you visit the page again with browser extensions disabled or in an incognito/private browsing window, or in another browser, do you experience the same trouble?
 
This can also sometimes occur when accessing GitHub using an outdated browser. Likewise, sometimes using a future Dev or Beta release can account for the same behavior. Browser versions update so frequently we recommend verifying that you're using the latest stable release of a supported browser.
 
With any luck, one of these options should resolve this issue, but please do let me know if you continue to run into any further trouble.
 
If you could confirm the exact browser version and operating system you're using as well as that would be most appreciated.
 
It would also be very helpful for us to see a screenshot of the Console view in your web browser.
 

    Google Chrome Console Overview
    Use the developer tools in the Develop menu in Safari on Mac
    Firefox Browser Console Overview

 
Hopefully there will be something in either the console or network logging that can help shed some light on this.
 
Once I have more information, I'll certainly do my best to help! 😁

All the best,
Rick

GitHub Support

**Comment by jonholdsworth** [comment]
Still happening.

Try opening the 300+ comments at the end in Linux chrome.

https://github.com/oven-sh/WebKit/pull/249

I've replied to their ticket with that url.

**Comment by mmDevel** [comment]
Yep this is still unusable, when I do a review it is fine, but open a PR or read a long file and Microsoft Edge is dead, with the whole computer. And I cannot get more than 32GB RAM, I am already at the limit by the company, so please fix it. Today it was already bad on 17 lines, fair that I had less resources available, but I just need to open the PR, and than I will review...

You can do it for review, than also do it for the rest. 

I do not mind using another browser just tell me which I will happily ditch Microsoft Edge, if it works in others.

**Comment by jonholdsworth** [comment]
OK if it's still happening, would everyone here mind filing a ticket with them?
If they get enough duplicates they'll have to take it seriously.
The problem is that it's a WorksForMe bug, and they are often the last to be actioned.

**Comment by tronicboy1** [comment]
Man I still have absolute horrible performance in GitHub pull requests (M2 Mac). Not even big PRs. Entering comments slows to a crawl. It has to be the lovely React rewrite re calculating the whole page on every key stroke. Please put back the old web components/vanilla JS backend based version.
